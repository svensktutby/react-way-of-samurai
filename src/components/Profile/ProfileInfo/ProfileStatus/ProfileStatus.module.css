.statusWrapper,
.editStatusWrapper {
  max-width: 300px;
}

.status {
  display: block;
  width: 100%;
  height: calc(1.5em + 0.75rem + 2px);
  padding: 0.375rem 0.75rem;

  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;

  box-shadow: 0 0.2rem 0.2rem rgba(0,123,255,0.25);
}

.editStatus {
  border-top: 0;
  border-right: 0;
  border-left: 0;
}

.editStatus:focus {
  box-shadow: 0 0.2rem 0.2rem rgba(0,123,255,0.25);
}
