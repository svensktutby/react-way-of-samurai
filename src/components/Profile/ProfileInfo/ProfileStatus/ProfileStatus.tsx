import React, { ChangeEvent, Component } from 'react';

import styleInput from '../../../common/styles/Input.module.css';
import s from './ProfileStatus.module.css';

type PropsType = {
  status: string;
  updateStatus: (status: string) => void;
};

type StateType = {
  status: string;
  editMode: boolean;
};

export class ProfileStatus extends Component<PropsType, StateType> {
  state = {
    status: this.props.status,
    editMode: false,
  };

  componentDidUpdate(prevProps: Readonly<PropsType>): void {
    this.updateStatusIfNeeded(prevProps.status);
  }

  activateEditMode = (): void => {
    this.setState({
      editMode: true,
    });
  };

  deactivateEditMode = (): void => {
    this.setState({
      editMode: false,
    });

    this.props.updateStatus(this.state.status);
  };

  statusChangeHandler = (e: ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      status: e.currentTarget.value,
    });
  };

  updateStatusIfNeeded(lastStatus: string): void {
    if (lastStatus !== this.props.status) {
      this.setState({
        status: this.props.status,
      });
    }
  }

  render(): JSX.Element {
    return (
      <>
        {!this.state.editMode ? (
          <div className={s.statusWrapper}>
            <span className={s.status} onDoubleClick={this.activateEditMode}>
              {this.props.status || 'No status'}
            </span>
          </div>
        ) : (
          <div className={`${styleInput.inputWrapper} ${s.editStatusWrapper}`}>
            <input
              className={`${styleInput.input} ${s.editStatus}`}
              value={this.state.status}
              onBlur={this.deactivateEditMode}
              onChange={this.statusChangeHandler}
              autoFocus
            />
          </div>
        )}
      </>
    );
  }
}
