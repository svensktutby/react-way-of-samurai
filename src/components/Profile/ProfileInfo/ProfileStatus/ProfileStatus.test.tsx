import React from 'react';
import TestRenderer, { ReactTestInstance } from 'react-test-renderer';

import { ProfileStatus } from './ProfileStatus';

let root: ReactTestInstance;
const mockCallback = jest.fn();
const props = { status: 'Status is not empty', updateStatus: mockCallback };

describe('ProfileStatus component', () => {
  beforeEach(() => {
    root = TestRenderer.create(<ProfileStatus {...props} />).root;
  });

  it('should matches the snapshot', () => {
    const component = TestRenderer.create(<ProfileStatus {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('status from props should be in the state', () => {
    const { instance } = root;
    expect(instance.state.status).toBe('Status is not empty');
  });

  it('after creation <span> should be displayed', () => {
    expect(() => {
      root.findByType('span');
    }).not.toThrow();
  });

  it("after creation <input> shouldn't be displayed", () => {
    expect(() => {
      root.findByType('input');
    }).toThrow();
  });

  it('after creation <span> should contains correct status', () => {
    const span = root.findByType('span');
    expect(span.children[0]).toBe('Status is not empty');
  });

  it('input should be displayed in editMode instead of span', () => {
    const span = root.findByType('span');
    span.props.onClick();

    const input = root.findByType('input');
    expect(input.props.value).toBe('Status is not empty');
  });

  it('callback should be called', () => {
    const { instance } = root;
    instance.deactivateEditMode();
    expect(mockCallback.mock.calls).toHaveLength(1);
  });
});
