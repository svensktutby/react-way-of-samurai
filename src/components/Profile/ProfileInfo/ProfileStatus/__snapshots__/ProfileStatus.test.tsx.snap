// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileStatus component should matches the snapshot 1`] = `
<div
  className="statusWrapper"
>
  <span
    className="status"
    onClick={[Function]}
    role="presentation"
  >
    Status is not empty
  </span>
</div>
`;
